core-ui/src/main/java/com/anytypeio/anytype/core_ui/features/editor/holders/media/File.kt

```diff
@@ -42,40 +44,58 @@ class File(val binding: ItemBlockFileBinding) : Media(binding.root), Decoratable
+
+            val displayMetrics: DisplayMetrics = itemView.context.resources.displayMetrics
+            val screenWidthPx = displayMetrics.widthPixels
+
+            val layoutParams = name.layoutParams as ViewGroup.MarginLayoutParams
+            val margins = layoutParams.marginStart + layoutParams.marginEnd
+            val paddings = name.paddingStart + name.paddingEnd
+
+            val textPaint = name.paint
+            val availableWidthPx = screenWidthPx - icon.width - margins - paddings
+            val maxTextWidthPx = availableWidthPx - textPaint.measureText("$fileExt $size")
+
+            val displayFileName =
+                if (textPaint.measureText(fileName + fileExt + size) > maxTextWidthPx) {
+                    val ellipsis = "..."
+                    val keepLength =
+                        (maxTextWidthPx - textPaint.measureText(ellipsis + fileExt + size)).toInt()
+                    TextUtils.ellipsize(
+                        fileName,
+                        textPaint,
+                        keepLength.toFloat(),
+                        TextUtils.TruncateAt.MIDDLE
+                    ).toString()
+                } else {
+                    fileName
+                }
 
-            val spannable = if (item.fileExt.isNullOrBlank()) {
-                SpannableString("${item.name}  $size")
-            } else {
-                SpannableString("${item.name}.${item.fileExt}  $size")
-            }
-            val start = if (item.fileExt.isNullOrBlank()) {
-                item.name!!.length + 2
-            } else {
-                item.name!!.length + item.fileExt!!.length + 2
-            }
-            val end = if (item.fileExt.isNullOrBlank()) {
-                item.name!!.length + 2 + size.length
-            } else {
-                item.name!!.length + item.fileExt!!.length + 3 + size.length
-            }
+            val fullText = "$displayFileName$fileExt $size"
+
+            val spannable = SpannableString(fullText)
+
+            val sizeStart = fullText.length - size.length
             spannable.setSpan(
                 RelativeSizeSpan(0.87f),
-                start,
-                end,
-                Markup.DEFAULT_SPANNABLE_FLAG
+                sizeStart,
+                fullText.length,
+                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
             )
             spannable.setSpan(
                 ForegroundColorSpan(itemView.context.color(R.color.text_secondary)),
-                start,
-                end,
-                Markup.DEFAULT_SPANNABLE_FLAG
+                sizeStart,
+                fullText.length,
+                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
             )
-            name.setText(spannable, TextView.BufferType.SPANNABLE)
+            name.text = spannable
         } else {
-            name.setText(item.name, TextView.BufferType.SPANNABLE)
+            name.text = item.name
         }
 
         applySearchHighlight(item)
```

Comment:
consider testing the following corner-case scenario: when searching inside the object (go to three-dots menu, then select "search", then enter search query containing characters from the original filename), what happens if the search highlight is applied to the ellipsized text sequence? there might be an issue with the span ranges here, because these text ranges are applied to the original filename.